src/pd/retry.rs

```diff
@@ -108,24 +115,84 @@ impl fmt::Debug for RetryClient {
+struct ShouldReconnect<'a> {
+    connected: MutexLockFuture<'a, bool>,
+    first_ready: Option<bool>,
+}
+
+impl<'a> ShouldReconnect<'a> {
+    fn new(connected: MutexLockFuture<'a, bool>) -> Self {
+        ShouldReconnect {
+            connected,
+            first_ready: None,
+        }
+    }
+}
+
+impl<'a> Future for ShouldReconnect<'a> {
+    // `tuple.0 : MutexGuard` tells whether current connecting is finished, `tuple.1 : bool` tells if
+    // we should start a new connection.
+    type Output = (MutexGuard<'a, bool>, bool);
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match Pin::new(&mut self.connected).poll(cx) {
+            Poll::Ready(val) => match self.first_ready {
+                None => {
+                    self.first_ready = Some(true);
+                    Poll::Ready((val, true))
+                }
+                Some(true) => panic!("should not happen"),
+                Some(false) => {
+                    let res = !*val;
+                    Poll::Ready((val, res))
+                }
+            },
+            Poll::Pending => {
+                if self.first_ready.is_none() {
+                    self.first_ready = Some(false)
+                }
+                Poll::Pending
+            }
+        }
+    }
+}
+
 // A node-like thing that can be connected to.
+#[async_trait]
 trait Reconnect {
     type Cl;
-    fn reconnect(&self, interval: u64) -> Result<()>;
+    async fn reconnect(&self, interval: u64) -> Result<()>;
     fn with_cluster<T, F: Fn(&Self::Cl) -> T>(&self, f: F) -> T;
 }
 
+#[async_trait]
 impl Reconnect for RetryClient<Cluster> {
     type Cl = Cluster;
 
-    fn reconnect(&self, interval: u64) -> Result<()> {
-        if let Some(cluster) =
-            self.connection
-                .reconnect(&self.cluster.read().unwrap(), interval, self.timeout)?
-        {
-            *self.cluster.write().unwrap() = cluster;
+    async fn reconnect(&self, interval: u64) -> Result<()> {
+        let connected: MutexLockFuture<bool> = self.connected.lock();
+        let (mut connected, should_connect) = ShouldReconnect::new(connected).await;
+        // if try to reconnect while a previous reconnect is running,
+        // the second reconnect will succeed in direct if the first one is succeed
+        if should_connect {
+            *connected = false;
+            if let Some(cluster) = {
+                let (id, members) = {
+                    let read_guard = self.cluster.read().unwrap();
+                    (read_guard.id, read_guard.members.clone())
+                };
+
+                self.connection
+                    .reconnect(id, &members, interval, self.timeout)
+                    .await?
+            } {
+                *self.cluster.write().unwrap() = cluster;
+            }
+            *connected = true;
+            Ok(())
+        } else {
+            Ok(())
```

Comment:
I may be very wrong, but rather than having this logic plus a custom future, could you take the write lock on `self.cluster` on entry to this method and then check it to see if connected? Then reconnect if necessary. You are guaranteed only one task will try to reconnect at a time by the RwLock, and when that task does complete, the other task can just return.

Reply:
> I may be very wrong, but rather than having this logic plus a custom future, could you take the write lock on `self.cluster` on entry to this method and then check it to see if connected? Then reconnect if necessary. You are guaranteed only one task will try to reconnect at a time by the RwLock, and when that task does complete, the other task can just return.

I think @nrc is right,maybe we over engineered in the current version, what's you opinion? @sticnarf